src/lib/topology.c

```diff
@@ -0,0 +1,1049 @@
+			trace_tplg_error("eut");
+			tplg.pos += hdr->payload_size;
+			break;
+		}
+	}
+	DEBUG_PRINT("topology parsing end\n");
+
+	for (i = 0; i < comp_num; i++)
+		rfree(comp_list[i].name);
+
+	rfree(comp_list);
+	return 0;
+}
+
+/* parse vendor tokens in topology */
+int sof_parse_tokens(void *object, const struct sof_topology_token *tokens,
+		     int count, struct snd_soc_tplg_vendor_array *array,
+		     int priv_size)
+{
+	int asize;
+
+	while (priv_size > 0) {
+		asize = array->size;
+
+		/* validate asize */
+		if (asize < 0) { /* FIXME: A zero-size array makes no sense */
+			DEBUG_PRINT("error: invalid array size 0x%x\n",
+				    asize);
+			return -EINVAL;
+		}
+
+		/* make sure there is enough data before parsing */
+		priv_size -= asize;
+		if (priv_size < 0) {
+			DEBUG_PRINT("error: invalid array size 0x%x\n",
+				    asize);
+			return -EINVAL;
+		}
+
+		/* call correct parser depending on type */
+		switch (array->type) {
+		case SND_SOC_TPLG_TUPLE_TYPE_UUID:
+			sof_parse_uuid_tokens(object, tokens, count, array);
+			break;
+		case SND_SOC_TPLG_TUPLE_TYPE_STRING:
+			sof_parse_string_tokens(object, tokens, count, array);
+			break;
+		case SND_SOC_TPLG_TUPLE_TYPE_BOOL:
+		case SND_SOC_TPLG_TUPLE_TYPE_BYTE:
+		case SND_SOC_TPLG_TUPLE_TYPE_WORD:
+		case SND_SOC_TPLG_TUPLE_TYPE_SHORT:
+			sof_parse_word_tokens(object, tokens, count, array);
+			break;
+		default:
+			DEBUG_PRINT("error: unknown token type %d\n",
+				    array->type);
+			return -EINVAL;
+		}
+
+		/* next array */
+		array = (void *)array + asize;
+	}
+	return 0;
+}
+
+void sof_parse_word_tokens(void *object,
+			   const struct sof_topology_token *tokens, int count,
+			   struct snd_soc_tplg_vendor_array *array)
+{
+	struct snd_soc_tplg_vendor_value_elem *elem;
+	int i, j;
+
+	/* parse element by element */
+	for (i = 0; i < array->num_elems; i++) {
+		elem = &array->value[i];
+
+		/* search for token */
+		for (j = 0; j < count; j++) {
+			/* match token type */
+			if (tokens[j].type != SND_SOC_TPLG_TUPLE_TYPE_WORD)
+				continue;
+
+			/* match token id */
+			if (tokens[j].token != elem->token)
+				continue;
+
+			/* matched - now load token */
+			tokens[j].get_token(elem, object, tokens[j].offset,
+				tokens[j].size);
+		}
+	}
+}
+
+void sof_parse_uuid_tokens(void *object,
+			   const struct sof_topology_token *tokens,
+			   int count, struct snd_soc_tplg_vendor_array *array)
+{
+	struct snd_soc_tplg_vendor_uuid_elem *elem;
+	int i, j;
+
+	/* parse element by element */
+	for (i = 0; i < array->num_elems; i++) {
+		elem = &array->uuid[i];
+
+		/* search for token */
+		for (j = 0; j < count; j++) {
+			/* match token type */
+			if (tokens[j].type != SND_SOC_TPLG_TUPLE_TYPE_UUID)
+				continue;
+
+			/* match token id */
+			if (tokens[j].token != elem->token)
+				continue;
+
+			/* matched - now load token */
+			tokens[j].get_token(elem, object, tokens[j].offset,
+				tokens[j].size);
+		}
+	}
+}
+
+void sof_parse_string_tokens(void *object,
```

Comment:
can we prefix all non static functions tplg_ instead of sof_

Reply:
OK, also there's a lot of function those can be static.